### PR TITLE
Sort clocks that will not expire

### PR DESCRIPTION
**Pull request by dexonsmith**
**Depends on #2442 — please merge that first.** (Does GitHub have a way to do stacked pull requests? The only commit in this PR *should be* fe286a71f49707f76f02411cdb50cb4ddbe338d4.)

Expose the `JGOFTimeControl` field in `GameType.json`, and use that to sort
clocks that aren't going to expire by computing the time remaining. This
includes:

- Sorting by the clock of the player whose turn it is NOT.
- Sorting by either clock in a paused game.

Note: 7ac13061552abd57827802f19bdde54c098a8308 fixed paused games to not be
intermingled with un-paused games, but was still sorting by an ill-defined
expiration time. This computes the time left on the clock and sorts by that
instead.

### CODE REVIEW

**Comment by dexonsmith** [comment]
Some screenshots follow.

Before, sorting by this clock:
before-this-clock

After, sorting by this clock:
after-this-clock

Before, sorting by opponent's clock:
before-opponent-clock

After, sorting by opponent's clock:
after-opponent-clock

**Comment by BHydden** [comment]
Just FYI, previously the clocks were all sorted by the running clock. Green games were sorted by your clock, white games were sorted by opponent's clock. The buttons at the top only toggled which group was on top and whether the sort was ascending or descending. 
Whether or not that system or yours is better is subjective, just wanted to make sure you knew what the old system was doing.
Thanks heaps for fixing the pause thing 👍

**Comment by dexonsmith** [comment]
> Just FYI, previously the clocks were all sorted by the running clock. Green games were sorted by your clock, white games were sorted by opponent's clock. The buttons at the top only toggled which group was on top and whether the sort was ascending or descending. Whether or not that system or yours is better is subjective, just wanted to make sure you knew what the old system was doing. Thanks heaps for fixing the pause thing 👍

Yes— in the old sort, `clock` and `opponent-clock` were both sorting by the currently-running clock. In the new sort, `clock` sorts by this player's clock and `opponent-clock` sorts by the opponent's clock. (Both before and after, `clock` puts "green" games first, and `opponent-clock` puts them last.)

IMO, the result with this patch is more intuitive... what I think a user might expect when they ask to sort by `clock` or `opponent-clock`.

I agree it's a bit subjective whether the new sort is strictly better, but on the upside there's no information loss. Although sorting by `clock` no longer shows "white" games in the order of when opponents' clocks run out, sorting by `opponent-clock` will still give that to you.

(That said, if I'm alone here and consensus is to leave un-paused games as they were, I could revert that part of the patch and only use "remaining" time to distinguish paused games.)

**Comment by dexonsmith** [comment]
I just did some further testing using `simple` time, and it doesn't work the way I thought— `black_time` and `white_time` are set to the same thing as `expiration` (not to time remaining). Should have a fix soon, but exhaustively testing the other clock types (such as absolute) to double check.

**Comment by dexonsmith** [comment]
> I just did some further testing using `simple` time, and it doesn't work the way I thought— `black_time` and `white_time` are set to the same thing as `expiration` (not to time remaining). Should have a fix soon, but exhaustively testing the other clock types (such as absolute) to double check.

Fixed in 7a5eb1a1ec52afc5f122856e7ad84fd5276c734a. Also pushed 6a5c42f to make the maybe-`undefined` nature of `time_control` explicit.

I tested all the time control types, and they sort relative to each other the way I expect:
- simple
- byoyomi
- fischer
- canadian
- absolute
- none

There are two wrinkles:
- `canadian` time remaining is calculated based on the full block time, regardless of how many moves are left in the block. Arguably that doesn't accurately reflect the game urgency. But I don't think this matters much (and Canadian time is an odd choice for online games anyway, since it's specifically designed for getting per-move time control using physical absolute-time chess clocks).
- `none` time remaining is calculated as "infinite"... which isn't actually true, but I think it's still reasonable to sort these last. The (60-day?) timeout gets reset when games are unpaused (maybe?) or any time a move is made (definitely!). And if sorting by the player's clock whose turn it is (and the game is not paused), `expiration` is used as usual, so no regression.

Anyway, the diff should be easier to look at after #2442 lands.

File: src/components/GameList/GameList.tsx [hunk 1]
```diff
@@ -80,7 +89,43 @@ export class GameList extends React.PureComponent<GameListProps, GameListState>
         }
     };
 
-    extractClockSortFields(game: GameType) {
+    isPaused(pause_control: AdHocPauseControl) {
```
[the remark this answers is not in the document]
Comment (dexonsmith) [reply src/components/GameList/GameList.tsx]: Done in a18157125121594835941e28cae3ea4ca05f50b7.

**Comment by anoek** [comment]
Thanks!

**Comment by dexonsmith** [comment]
> PS. If you are going to change it so that all games sort by either your clock or opponents clock, maybe you would also want to comingle the green and white groups? Just a thought.

Just seeing this comment: I think not, at least for now, because:
- IMO, it's useful that "my move" sorts separately from "not my move", and probably what most players want from a clock-sort. For `clock`, we're getting "order of urgency of my attention"; for `opponent-clock`, we get "order of urgency of my opponents' attention".
- It's nice/convenient to keep using `expiration` (timestamp game ends) when sorting by the player whose clock is running, and this only works if we separate green and white. One benefit of `expiration` is that it incorporates the timeout for games with time control `none`.

**Pull request merged**